lib/segment/src/payload_storage/mmap_payload_storage.rs

```diff
@@ -0,0 +1,172 @@
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+
+use blob_store::config::StorageOptions;
+use blob_store::{Blob, BlobStore};
+use common::types::PointOffsetType;
+use parking_lot::RwLock;
+use serde_json::Value;
+
+use crate::common::operation_error::{OperationError, OperationResult};
+use crate::common::Flusher;
+use crate::json_path::JsonPath;
+use crate::payload_storage::PayloadStorage;
+use crate::types::{Payload, PayloadKeyTypeRef};
+
+const STORAGE_PATH: &str = "payload_storage";
+
+impl Blob for Payload {
+    fn to_bytes(&self) -> Vec<u8> {
+        serde_json::to_vec(self).unwrap()
+    }
+
+    fn from_bytes(data: &[u8]) -> Self {
+        serde_json::from_slice(data).unwrap()
+    }
+}
+
+#[derive(Debug)]
+pub struct MmapPayloadStorage {
+    storage: Arc<RwLock<BlobStore<Payload>>>,
+}
+
+impl MmapPayloadStorage {
+    pub fn open_or_create(path: &Path) -> OperationResult<Self> {
+        let path = path.join(STORAGE_PATH);
+        if path.exists() {
+            Self::open(path)
+        } else {
+            // create folder if it does not exist
+            std::fs::create_dir_all(&path).map_err(|_| {
+                OperationError::service_error("Failed to create mmap payload storage directory")
+            })?;
+            Ok(Self::new(path))
+        }
+    }
+
+    fn open(path: PathBuf) -> OperationResult<Self> {
+        if let Some(storage) = BlobStore::open(path) {
```

Comment:
we should have an error here explaining exactly why it can't opened

Reply:
https://github.com/qdrant/qdrant/pull/5398/commits/8f697c26c0e14aa7c417c58c1bcee5e260f11431